### PR TITLE
Recover from `Foo(a: 1, b: 2)`

### LINKED ISSUE #61326: Handle `Foo { a: usize, b: usize }` used as `Foo(1, 2)` gracefully
Given the code

https://github.com/rust-lang/rust/blob/3592079765d4d76f19b3a9501de5f8cc47f11a04/src/test/ui/parser/recover-from-bad-variant.rs#L1-L14

we emit

https://github.com/rust-lang/rust/blob/3592079765d4d76f19b3a9501de5f8cc47f11a04/src/test/ui/parser/recover-from-bad-variant.stderr#L1-L21

We should emit something closer to

```
error: expected type, found `3`
  --> $DIR/recover-from-bad-variant.rs:7:26
   |
LL |     let x = Enum::Foo(a: 3, b: 4);
   |                          ^ expecting a type here because of type ascription

error[E0532]: expected tuple struct/variant, found struct variant `Enum::Foo`
  --> $DIR/recover-from-bad-variant.rs:10:9
   |
LL |         Enum::Foo(a, b) => {}
   |         ^^^^^^^^^ did you mean `Enum::Foo { a, b }`?

error: aborting due to 2 previous errors
```

### PR DESCRIPTION
**Pull request by estebank**
Detect likely `struct` literal using parentheses as delimiters and emit
targeted suggestion instead of type ascription parse error.

Fix #61326.

### CODE REVIEW

File: src/test/ui/parser/issue-44406.stderr [hunk 1]
```diff
@@ -9,17 +9,25 @@ help: you can escape reserved keywords to use them as identifiers
...
+   |         ^^^^^^^^^^^^^^^
 ...
 LL |     foo!(true);
-   |          ^^^^ expected type
+   |     ----------- in this macro invocation
+   |
+   = note: this error originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: if `bar` is a struct, use braces as delimiters
+   |
+LL |         bar {  }
```
Comment (notriddle) [inline src/test/ui/parser/issue-44406.stderr:24]: This seems messed up. Shouldn't it be `bar { baz: $rest }` ?

Comment (estebank) [reply src/test/ui/parser/issue-44406.stderr:24]: This is indeed "not correct", caused by the macro context of the spans involved. I could _never_ show the suggestions if a macro is involved, but that would hide the suggestion even if the macro _isn't_ mixing fragments coming from the caller with internal spans. Then again a middle point would be to give just a `help` without a suggestion for things in macros now that I think about it 🤔

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -1031,9 +1037,56 @@ impl<'a> Parser<'a> {
...
+            Some((self.clone(), fun.kind.clone()))
+        } else {
+            None
+        };
+        let open_paren = self.token.span;
+
+        let mut seq = self.parse_paren_expr_seq().map(|args| {
             self.mk_expr(lo.to(self.prev_token.span), self.mk_call(fun, args), AttrVec::new())
         });
+        match (seq.as_mut(), snapshot) {
```
Comment (oli-obk) [inline compiler/rustc_parse/src/parser/expr.rs:1052]: Please pull this out of the happy path.

r=me with that done

Comment (estebank) [reply compiler/rustc_parse/src/parser/expr.rs:1052]: very rude of you to make me fix this while Mara is speaking :)

**Pull request merged**